Client/Frontend/Home/TopSites/DataManagement/ContileProvider.swift

```diff
@@ -0,0 +1,89 @@
+//// This Source Code Form is subject to the terms of the Mozilla Public
+//// License, v. 2.0. If a copy of the MPL was not distributed with this
+//// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import Foundation
+import Shared
+
+typealias ContileResult = Swift.Result<[Contile], Error>
+
+protocol ContileProviderInterface {
+
+    /// Fetch contiles either from cache or backend
+    /// - Parameters:
+    ///   - timestamp: The timestamp to retrieve from cache, useful for tests. Default is Date.now()
+    ///   - completion: Returns an array of Contile, can be empty
+    func fetchContiles(timestamp: Timestamp, completion: @escaping (ContileResult) -> Void)
+}
+
+extension ContileProviderInterface {
+    func fetchContiles(timestamp: Timestamp = Date.now(), completion: @escaping (ContileResult) -> Void) {
+        fetchContiles(timestamp: timestamp, completion: completion)
+    }
+}
+
+/// `Contile` is short for contextual tiles. This provider returns data that is used in Shortcuts (Top Sites) section on the Firefox home page.
+class ContileProvider: ContileProviderInterface, Loggable, URLCaching {
+
+    static let contileResourceEndpoint = "https://contile.services.mozilla.com/v1/tiles"
+
+    lazy var urlSession = makeURLSession(userAgent: UserAgent.mobileUserAgent(),
+                                         configuration: URLSessionConfiguration.default)
+
+    enum Error: Swift.Error {
+        case failure
```

Comment:
Should this be clearer? Should it have multiple cases of failure type that we're exposing?

Reply:
For parent manager it's either we have data from the provider or not. Name can change but I don't see how other cases will be used.